### DISCUSSION TITLE
[V3] - CSP Header

### DISCUSSION THREAD

**Discussion by mboelling** [livewire/livewire#6113] [1. Help]
Will the new version V3 of livewire be compatible to CSP Headers. I can't get V2 working with CSP Headers. 
If tried implementing it with nginx and with laravel-spatie package with no success, because violating the unsafe-eval directive.

**Comment by piotrku** [comment]
Using the latest Livewire 3.3.5.
Some errors were solved but this 

```
livewire.js Uncaught EvalError: Refused to evaluate a string as JavaScript because 'unsafe-eval' is not an allowed source of script in the following Content Security Policy directive: "script-src 'self' 'nonce-2CyoSDLv7FRdNs1uxEuoFKM95gvVtxTqt3vaOUUy' ....
```

is still present, and this points to this place of `livewire.js`:

```
    const safeAsyncFunction = () => {
      try {
        let func2 = new AsyncFunction(["__self", "scope"], `with (scope) { __self.result = ${rightSideSafeExpression} }; __self.finished = true; return __self.result;`);
        Object.defineProperty(func2, "name", {
          value: `[Alpine] ${expression}`
        });
        return func2;
      } catch (error2) {
        handleError(error2, el, expression);
        return Promise.resolve();
      }
    };
```

**Reply by popchenk** [reply]
any update on this ? I am facing the same issue.

Thank you

**Comment by kemalyen** [comment]
I also need an urgent update, otherwise have to re-write anything because the pentest team is not approving the newly built application

**Comment by samods2** [comment]
Any update on this?

**Comment by NullPointerReference** [comment]
As a Security Engineer I would strongly suggest addressing this, as 'unsafe-eval' is considered problematic in most security-sensitive contexts and a working solution seems feasible now.

**Comment by blakedeckard** [comment]
Edit: This is WIP, as there is different behavior on Dev with Herd and in PRD, feel free to deploy and try and let me know. I will update as I alter things, feel free to add or suggest any different approaches.

I found the following solution, but it requires sending livewire/alpine on each request, including on wire:navigate. I'm wondering if there's a better way to do this using a hash as opposed to a nonce, but I am learning about the process as I go. For my simple app, this may do.

1. In each layout generate a nonce and pass it to your CSP script-src, style-src, and ensure connect-src has ws:
`

<?php $nonce = base64_encode(random_bytes(32)); ?>
                <meta name="Content-Security-Policy" content="default-src 'self';
                script-src 'self' 'nonce-{$nonce}';
                style-src 'self' 'nonce-{$nonce}';
                img-src 'self;
                media-src 'self';
                font-src 'self' https://fonts.bunny.net;
                connect-src 'self' ws:;
                upgrade-insecure-requests;">
`
I found if you try and do this through a partial and @include it will fail, as it will not get the php variable later in the body. It worked in dev with Herd, but broke in PRD.

2. In the body, as it is reloaded in each request, pass the livewireScripts with the nonce. If we place it in the header, there will be duplicates, as a new nonce is generated on each request, but only the body tags are swapped out by Livewire: https://livewire.laravel.com/docs/navigate.

In `app.blade.php`
`

        <body>
        @livewireStyles($nonce)
        @livewireScripts($nonce)

`


3. Publish the Livewire config, and set inject assets to false, this prevents duplicates from being generated as Livewire is typically injected without the nonce.
`php artisan livewire:publish --config`
`"inject_assets" => false,`

**Reply by blakedeckard** [reply]
I found moving @livewireStyles($nonce) and @livewireScripts($nonce) to the body works if you want to continue to use wire:navigate, as the body is replaced on each request.

I found that I was wrong in my original reply in stating that it only affected going between layouts as I found issues between pages using wire:navigate on my app layout. Thus, the above approach seems better, albeit, costly from a network perspective. I am editing the above post now.

**Comment by dorkster100** [comment]
There are several issues with csp and wire:navigate:
1) is related to https://github.com/livewire/livewire/commit/c636ba6fb92863c2b5535c8e04a683662ee4ab1d
it probably is needed to add `nonce` to `attributesExemptFromScriptTagHashing` array
2) when `navigate` happens the page is not reloaded, thus the headers are still having old csp_nonce, new nonce that is used by returned styles and scripts is not valid.

**Reply by dorkster100** [reply]
actually second point probably is wrong, since the initial script that inserts the scripts has correct nonce, it does not matter what nonce the child script has? But the styles are different, there is no correct way to append styles correctly? https://github.com/w3c/webappsec-csp/issues/625

**Comment by severfire** [comment]
Up :-/

**Comment by Fimazzu** [comment]
esto ha sido una porquería, parece que csp actualizo y dejo livewire inservible ahora debo reescribir demasiado codigo a fecha de hoy ! 19/12/2014

**Comment by incoming-th** [comment]
Having this issue too.

Using 

```
@livewireScripts(['nonce' => csp_nonce('script')])
```

in blade template, and

```
'inject_assets' => false,
```

in config file, the nonce is added correctly when the page load (no error), but when the page has a **wire:poll.10s="refresh"** for example, the below error is thrown:

```
Alpine Expression Error: Refused to evaluate a string as JavaScript because 'unsafe-eval' is not an allowed source of script in the following Content Security Policy directive: "script-src 'self' 'nonce-MDA4ZDlhZjZhOTUxYmZkMg=='

Expression: "$wire.refresh"

Uncaught EvalError: Refused to evaluate a string as JavaScript because 'unsafe-eval' is not an allowed source of script in the following Content Security Policy directive: "script-src 'self' 'nonce-MDA4ZDlhZjZhOTUxYmZkMg==' 
```

Could be because the nonce is not passed to Alpine?

**Reply by souhayla-dvp** [reply]
Hey,
Have you found a solution for this??

**Reply by incoming-th** [reply]
@souhayla-dvp Unfortunately no, I had to change my CSP to make alpine and livewire works. More headache during the pentest and OSWASP to justify this.

**Comment by binaryfire** [comment]
Does anyone know if there are plans to fix this? There are 3 open discussions dating back 3 years and no official response yet:

https://github.com/livewire/livewire/discussions/4884
https://github.com/livewire/livewire/discussions/6468
https://github.com/livewire/livewire/discussions/6113

**Comment by romatexdev** [comment]
Is there any news on this issue?

**Comment by oOMrCookieOo** [comment]
This did it for me
adding this to the nginx config
Am using docker nginx in the server (etc/nginx/sites-av) and the nginx image for my app
added it both configs

add_header Content-Security-Policy "default-src 'self' http: https: data: blob: 'unsafe-inline' 'unsafe-eval'; script-src 'self' 'unsafe-inline' 'unsafe-eval' https://cdn.jsdelivr.net https://unpkg.com; style-src 'self' 'unsafe-inline' https://fonts.googleapis.com https://fonts.bunny.net https://cdn.jsdelivr.net; font-src 'self' https://fonts.gstatic.com https://fonts.bunny.net data:; img-src 'self' data: https: blob:; connect-src 'self' ws: wss: https:; worker-src 'self' blob:;" always;

**Reply by NullPointerReference** [reply]
This is NOT a solution, this is enabling an insecure CSP setting as a workaround. It is 2025 and NOBODY should be using anything 'unsafe-<whatever>' in their CSP anymore. The solution is to properly refactor the livewire code so it does not require unsafe settings.

**Reply by oOMrCookieOo** [reply]
I agree. But cant really do much about it, livewire is already baked-into filament. in my usecase it's not something sensetive just simple cruds i dont have the time to change stack or create them manually

**Reply by NullPointerReference** [reply]
It'sgood to hear it is not something sensitive ! With modern EU regulations, one is always responsible for the tech-stack it uses and thus even when faced with legacy and baked sub-dependencies, one will still be held accountable, hence Livewire should really get with modern times; it's not even a very hard fix ..

**Reply by severfire** [reply]
I guess many ppl would even want to sponsor to fix it

**Reply by severfire** [reply]
i wonder how much ai tokens would take to fix this issue

**Reply by NullPointerReference** [reply]
Good one; it depends on the AI + model version I guess :)
I have no more 'stake' in this, as it was the reason to stop using it, but I think fixing it is very doable, with or without AI.

**Reply by severfire** [reply]
if you would write prompt to fix it, what would you write? :->

**Reply by NullPointerReference** [reply]
Well, I would need to dig into the issue in detail again, it's been 2 years. So I asked an AI instead ... even without using a token; as asking it to do all the refactoring would require some more significant time effort from my side; but I think the initial response looks very promising to have a code-generation/refactoring prompt fix these "quick-wins" automatically; so one can focus on the few problematic functions - if you even use that functionality.   

https://grok.com/share/bGVnYWN5LWNvcHk%3D_14c141d9-dbc9-43d5-a9e9-6e6f298c1d83

Disclaimer: Please do note that this was an initial prompt just to get a feeling for its suggestions and proposed solutions/approach; it can definitely use some refinement beyond the initial iteration.

**Comment by severfire** [comment]
it would be great to do fork or fix for it

**Comment by ghabriel25** [comment]
Its already has fix commit https://github.com/livewire/livewire/pull/8793 . Hope it will be available on V4

**Reply by severfire** [reply]
so, when i update my scripts and filament this may be resolved  / fixed? cool!

**Comment by alxndrmlr** [comment]
FYI folks 

https://github.com/alpinejs/alpine/pull/4671
https://github.com/livewire/livewire/pull/9461

**Reply by severfire** [reply]
WOW, ubercool!
